### PR TITLE
Remove final in protocol extensions

### PR DESCRIPTION
**Pull request by KingOfBrian**
This PR generates an error when the `final` attribute is added to a function or variable inside of a protocol extension. This matches the current semantic model -- `final` only disables overrides for class types, and it has no effect on protocol extensions. This makes `final` syntax consistent across struct, enum and protocol types.

This PR also cleans up the reporting of the message by adding a fixit, and only reporting the error once in variable declarations with child declarations. For instance, the error message is currently reported multiple times if a final variable declaration has implicit getters and setters.

This error message also uncovered a lot of usage in the test suite, and a few places where `final` was used in extensions of a Error protocols in Foundation.

Resolves [SR-1762](https://bugs.swift.org/browse/SR-1762).

### CODE REVIEW

**Comment by slavapestov** [comment]
Can you make it a warning in Swift 3 mode and add a testcase to test/Compatibility/?

**Comment by KingOfBrian** [comment]
It is a warning in Swift 3 and I added `test/Compatibility/final_protocol_extension.swift` which checks the warning in Swift 3 mode. Did I miss a case there?

**Comment by slavapestov** [comment]
Oh sorry I missed that you already did that. Looks good, thanks!

**Comment by slavapestov** [comment]
@jrose-apple Does this look ok to you?

**Comment by slavapestov** [comment]
@swift-ci Please smoke test

**Comment by jrose-apple** [comment]
I mentioned on the bug (and Joe concurred) that this should still go through swift-evolution, since it's an additional break between Swift 3 and Swift 4. It's tiny but that doesn't automatically mean we should just do it.

File: test/attr/attr_final.swift [hunk 1]
```diff
@@ -1,4 +1,4 @@
-// RUN: %target-typecheck-verify-swift
+// RUN: %target-typecheck-verify-swift -swift-version 4
```
Comment (jrose-apple) [inline test/attr/attr_final.swift]: This is not a change we should make just yet. We're using our existing test suite as source compatibility tests.

**Comment by jrose-apple** [comment]
We can certainly put the "cleanup" changes in to remove final where it's not necessary, and we can add the warning unconditionally for now.

File: test/attr/attr_final.swift (hunk 1 above)
Comment (KingOfBrian) [inline test/attr/attr_final.swift]: Are you referring to the `-swift-version 4` or to the changes to the tests themselves? I thought that all of the source compatibility tests (for v3) were being put int `test/Compatibility`. I added `-swift-version 4` here in an attempt to follow the pattern I saw in `test/attr_[escaping|inlineable|objc]`, but glad to change anything as needed.

Comment (jrose-apple) [inline test/attr/attr_final.swift]: `-swift-version 4`. You only get to add that if *everything* in the file is also being tested in Compatibility. (Right, @slavapestov?)

**Comment by KingOfBrian** [comment]
The `attr->invalidate()` that I added to remove duplicate warnings was causing the ASTVerifier to fail when comparing the func decl final state to the storage decl final state. Instead of invalidating the attribute, I am removing it which should make sure that the func and storage final states are consistent, and that the error message only occurs once.

I also made a PR to foundation(apple/swift-corelibs-foundation#917) to remove the two places where this warning would occur.

**Comment by alblue** [comment]
@swift-ci please smoke test

**Review by jrose-apple — APPROVED** [review]

**Comment by jrose-apple** [comment]
Thanks, Brian!

**Pull request merged**